### ISSUE TITLE
Can `get_config()` be cached/memoized?

### ISSUE THREAD

**Issue by ronawho** [Bears-R-Us/arkouda#871]
Currently `get_config()` always results in a server call, but the server config doesn't change so I think we could cache the results, at least for the current implementation. However, I'm not sure if this could create problems in the future so I wanted to open this issue to discuss.

**Comment by glitch** [comment]
I looked through the client source code for `get_config()` and it's only really used in the following:

- `benchmarks` (13 times and in each case it is only once per file)
- `tests` (6 times, the `join_test.py` calls it 4 times which can probably be cleaned up).

Looking through the unit test usage we can clean up the 4 calls in `join_test`.

Looking at the benchmark testing, the get_config call is being invoked at the start of a function directly invoked by `__main__` where a fresh arkouda client is being stood up so I don't think caching is going to make a difference (a new python interpreter is being used each time).

Here's an example:  [benchmarks/IO.py](https://github.com/Bears-R-Us/arkouda/blob/master/benchmarks/IO.py)

main is called, client is created / connects to server, `time_ak_write_read` is called in main and calls `get_config` to retrieve the number of locales.  This is the pattern across all 13 of the benchmark cases.

**Comment by ronawho** [comment]
Should have mentioned this, but I'm looking to add byteorder/endianness to get_config and I'll need to query endianness for each to_ndarray/ak.array() call and I was wondering if I should just cache endianness or if the entire server config should be cached.

e.g. what my current branch does is:

```python
@functools.lru_cache(maxsize=None)
def get_server_byteorder() -> str:
    from arkouda.client import get_config
    order = get_config()['byteorder']
    return cast('str', order)
```

and I was wondering if this caching should be moved up to the get_config level instead.

**Comment by mhmerrill** [comment]
@ronawho We could cache a new copy each time we connect to the server.

**Comment by glitch** [comment]
@ronawho  I put up [PR#873](https://github.com/Bears-R-Us/arkouda/pull/873) to implement the `@lru_cache` on get_config but also added in a default time expiration of 1 hour (you can see stack overflow post/answer https://stackoverflow.com/a/55900800 for the explanation).

This also returns the config object as part of the `ak.connect` call.  Let me know what you think.

**Comment by ronawho** [comment]
Great, thanks! What's the motivation for the 1 hour expiration?

**Comment by glitch** [comment]
> Great, thanks! What's the motivation for the 1 hour expiration?

It's more of a future fail-safe than anything else.  I know there was another issue talking about auto-tuning some parameters at run-time and dynamic adjustment probably wouldn't be too far off from there.

I don't know, maybe we don't need the time component at all.  It could be I'm nervous from prior experience on other projects where caching these turned out to byte us :)

To be honest, I'm not certain this should even be cached.  I get that unit tests and benchmarks might call it repeatedly, but maybe those tests should be started with a known configuration or store the configuration values themselves after startup.  I don't think this is something that users are ever really going to do.

Then again, there really isn't much of a compelling reason NOT to cache it either, so the time thing felt like an in-between.

If you feel like the implementation should be different, we can always decline the PR, I'm not really invested with it either way.

Cheers!

**Comment by ronawho** [comment]
> I don't know, maybe we don't need the time component at all. It could be I'm nervous from prior experience on other projects where caching things turned out to bite us :)

I share the general concern about caching the configuration and having somebody in the future add something that's not cacheable to the config. Would making the config that lives in ServerConfig made `const` alleviate that concern for you?

For the 1 hour expiration, my concern was that if there are dynamic adjustments the 1-hour cache could make any issues sporadic and harder to track down. 

> To be honest, I'm not certain this should even be cached. I get that unit tests and benchmarks might call it repeatedly, but maybe those tests should be started with a known configuration or store the configuration values themselves after startup. I don't think this is something that users are ever really going to do.

I agree this is unlikely something a user will do. My use-case is for optimizing `to_ndarray/ak.array()`. I need to know the byteOrder of the server, but I don't want to make an additional server call each time to figure this out.

**Comment by glitch** [comment]
Ah... I think I understand where you're going now.  If I understand correctly this isn't really an issue about caching and updates as much as an issue that we aren't storing the config info to begin with.  You need a value and you're forced to call get config to go and retrieve it because the original result is not centrally available.  Sorry, I think I read my own bias / connotation with the word _cache_ into this which to me means we want to temporarily store something that could be updated over time.

On that note,  I agree, saving the config info in an object on the client side upon start-up makes sense.  Basically get the "running config".

Sorry about that :)
